tests/integration/HACKING.md

```diff
@@ -0,0 +1,468 @@
+
+1. Creating tasks
+2. Creating a plugin class 
+3. Adding a girder\_worker\_plugins [entry point](http://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins) to your setup.py.
+
+Creating tasks follows the [standard celery conventions](http://docs.celeryproject.org/en/latest/userguide/tasks.html#basics). The only difference is the celery application that decorates the function should be imported from girder_worker.app. E.g.:
+
+```python
+from girder_worker.app import app
+
+@app.task
+def fibonacci(n):
+    if n == 1 or n == 2:
+        return 1
+    return fibonacci(n-1) + fibonacci(n-2)
+	
+```
+
+Each plugin must define a plugin class the inherits from ```girder_worker.GirderWorkerPluginABC```. GirderWorkerPluginABC's interface is simple. The class must define an \_\_init\_\_ function and a task\_imports function. \_\_init\_\_ takes the girder\_worker's celery application as its first argument. This allows the plugin to store a reference to the application, or change configurations of the application as necessary. The task\_imports function takes no arguments and must return a list of the package paths (e.g. importable strings) that contain the plugin’s tasks. As an example:
+
+```python
+from girder_worker import GirderWorkerPluginABC
+
+class GWExamplePlugin(GirderWorkerPluginABC):
+    def __init__(self, app, *args, **kwargs):
+        self.app = app
+
+        # Update the celery application's configuration
+        # it is not necessary to change the application configuration
+        # this is simply included to illustrate that it is possible.
+        self.app.config.update({
+            'TASK_TIME_LIMIT': 300
+        })
+
+    def task_imports(self):
+        return ['gwexample.analyses.tasks']
+		
+```
+
+Finally, in order to make the plugin class discoverable, each plugin must define a custom entry point in its setup.py. For our example, this entry point looks like this:
+
+```python
+from setuptools import setup
+
+setup(name='gwexample',
+      # ....
+      entry_points={
+          'girder_worker_plugins': [
+              'gwexample = gwexample:GWExamplePlugin',
+          ]
+      },
+      # ....
+      )
+```
+
+Python Entry Points are a way for python packages to advertise classes and objects to other installed packages. Entry points are defined in the following way:
+
+```python
+entry_points={
+    'entry_point_group_id': [
+        'entry_point_name = importable.package.name:class_or_object',
+    ]
+}
+```
+
+The girder\_worker package introduces a new entry point group girder\_worker\_plugins. This is followed by a list of strings which are parsed by setuptools. The strings must be in the form name = module:plugin\_class Where name is an arbitrary string (by convention the name of the plugin), module is the importable path to the module containing the plugin class, and plugin_class is a class that inherits from GirderWorkerPluginABC.
```

Comment:
would be good to say when it is parsed by setuptools, i.e. `are parsed by setuptools at the time the package is pip installed`. or whatever is the right thing to say.